weasyl/test/test_login.py

```diff
@@ -0,0 +1,524 @@
+# This Python file uses the following encoding: utf-8
+import unittest
+import pytest
+import py
+import bcrypt
+import arrow
+import web
+import os
+import json
+
+import db_utils
+from weasyl import login, macro
+from weasyl import define as d
+from weasyl.error import WeasylError
+from _pytest import tmpdir
+
+# Main test account
+raw_password = "0123456789"
+bcrypt_hash = login.passhash(raw_password)
+old_2a_bcrypt_hash = "$2a$12$qReI924/8pAsoHu6aRTX2ejyujAZ/9FiOOtrjczBIwf8wqXAJ22N."
+
+TestAccountName_nameprefix = "testlogin"
+TestAccountName_counter = 0
+
+class TestFunctions():
+    def generateTestAccountName(self):
+        global TestAccountName_counter
+        global TestAccountName_nameprefix
+        ret = TestAccountName_nameprefix + str(TestAccountName_counter)
+        TestAccountName_counter += 1
+        return ret
+    def generateTokenString(self, size):
+        import random
+        import string
+        return "".join(random.SystemRandom().choice(string.ascii_letters + string.digits) for i in range(size))
+
+class Bag(object):
+    def __init__(self, **kw):
+        for kv in kw.items():
+            setattr(self, *kv)
+
+class SigninTestCase(unittest.TestCase):
+    def testVerifyLoginRecordIsUpdated(self):
+        user_name = TestFunctions().generateTestAccountName()
+        user_id = db_utils.create_user(username=user_name)
+        d.execute("UPDATE login SET last_login = '-1' WHERE userid = %i", [user_id])
+        # login.signin(user_id) -will- raise an AttributeError when d.web.ctx.weasyl_session
+        #   tries to execute itself; so catch/handle; it's a test environment issue
+        self.assertRaises(AttributeError, login.signin, user_id)
+        query = d.execute("SELECT last_login FROM login WHERE userid = %i", [user_id])
+        self.assertGreater(query[0][0], -1)
+
+class AuthenticateBcryptTestCase(unittest.TestCase):
+    def testNoUsernameProvided(self):
+        result = login.authenticate_bcrypt(username='', password='password')
+        self.assertEqual(result, (0, 'invalid'))
+
+    def testNoPasswordProvided(self):
+        user_name = TestFunctions().generateTestAccountName()
+        result = login.authenticate_bcrypt(username=user_name, password='')
+        self.assertEqual(result, (0, 'invalid'))
+
+    def testInvalidUsernameProvided(self):
+        user_name = TestFunctions().generateTestAccountName()
+        result = login.authenticate_bcrypt(username=user_name, password=raw_password)
+        self.assertEqual(result, (0, 'invalid'))
+
+    def testLoginFailsForModsWithInvalidAuthentication(self):
+        log_path = '%s%s.%s.log' % (macro.MACRO_SYS_LOG_PATH, 'login.fail', d.get_timestamp())
```

Comment:
I'm still not entirely sure how to use PyTest here to generate the temporary directory. I know that when in use it would look something like the following (and I think it would transit back through to let login.py use it), but I don't know how to get there.
```
macro_original_path = macro.MACRO_SYS_LOG_PATH
macro.MACRO_SYS_LOG_PATH = tempDirectory
[... much code later ...]
macro.MACRO_SYS_LOG_PATH = macro_original_path
```